### ISSUE TITLE
Undertow Support?

### ISSUE THREAD

**Issue by uromahn** [dropwizard/dropwizard#1463]
Just wanted to "re-open" the question which was posted in Issue #475.

**Comment by joschi** [comment]
@uromahn The reply in this issue still applies: https://github.com/dropwizard/dropwizard/issues/475#issuecomment-35629030

**Comment by uromahn** [comment]
I understand.
I was looking at the Dropwizard project and noticed how "deep" Jetty is integrated into it. A lot of features are actually based on some Jetty features or extensions. I personally don't like the close coupling between Jetty and the Dropwizard framework, but I guess there is no easy way out.
I may take a look to find out how much effort it would be to de-couple Dropwizard better from Jetty and how much effort it would be to replace it with an Undertow.
On the other hand, while Undertow seems significantly more efficient and "faster" than Jetty, the question is whether this is really relevant and the effort to support Undertow would be worth it.

**Comment by joschi** [comment]
@uromahn
> On the other hand, while Undertow seems significantly more efficient and "faster" than Jetty

Do you know of any benchmarks supporting this claim you'd like to share?

**Comment by uromahn** [comment]
I was looking at this: http://www.techempower.com/benchmarks/

However, I was first looking at an outdated benchmark from three years ago which compared Jetty 8 with Undertow. The most recent benchmark (round 11) now shows that Jetty 9 has closed the "gap".  Undertow is still pretty fast but Jetty is now listed slightly better (http://www.techempower.com/benchmarks/#section=data-r11&hw=ec2&test=json). However, it is also noticeable that the same request with a Servlet on Jetty is still about 1.75 times slower than Undertow or pure jetty. What is even more surprising to me is that Dropwizard is listed as 5x **slower** than pure Jetty or Undertow.

Not sure though if those "benchmarks" are even representative of and meaningful for anything since it seems to compare "apples with oranges". I think a "real" benchmark would be to compare Jetty-Servlet with Undertow-Servlet or Dropwizard with Jetty and Dropwizard with Undertow (which is unfortunately not possible).

**Comment by joschi** [comment]
@uromahn Looking at https://www.techempower.com/benchmarks/#section=data-r11&hw=ec2&test=json&s=2&p=5179c-0-0, Jetty seems to be faster than Undertow. This being said, Dropwizard is a complete application framework consisting of the underlying web server (Jetty), the low-level HTTP abstraction (Java Servlets), and the high-level HTTP abstraction (JAX-RS/Jersey). In order to compare Undertow with Jetty in the context of Dropwizard, you'd have to build a similar stack on Undertow and benchmark it.

tl;dr I don't think that swapping Jetty with Undertow in Dropwizard will provide any gains, either performance-wise or for user-friendliness.

**Comment by cricket007** [comment]
With latest updates on the json test, Jetty is higher than Dropwizard, Undertow is still above Jetty 

https://www.techempower.com/benchmarks/#section=data-r18&hw=ph&test=json&l=xan3h7-f

I wrote the [Armeria Dropwizard Bundle](https://github.com/line/armeria/blob/master/site/src/sphinx/advanced-dropwizard-integration.rst) 
to hand off the ServerConnector to a separate Jetty Server (which I think is handled by Netty)

So, in theory anything with some bridge to Jetty should work, such as [Jooby](https://jooby.io/#server), so I think the question should become - why limit yourself to Jetty? Jooby and Spring offer different embedded servers, so why can't Dropwizard?

**Comment by j4n1** [comment]
> so I think the question should become - why limit yourself to Jetty?
What's wrong with Jetty? Jetty is mature & open source, well maintained and isn't going anywhere.
Jetty is fast, are there faster servers? sure, are those benchmarks relevant to any real world scenario, not at all.
Adding support for more servers means more code, consequently more bugs, worse out-of-the-box configuration and adds up a lot to the maintenance, adding no value.
Dropwizard is opinionated and it should stay opinionated. 
Want bloated framework? Spring is the way to go.

**Comment by joschi** [comment]
> With latest updates on the json test, Jetty is higher than Dropwizard, Undertow is still above Jetty

And Dropwizard (71), which is basically Jetty + Jersey, is faster than Undertow+Jetty (74) in the very benchmark you've linked.


> I think the question should become - why limit yourself to Jetty?

To quote the first sentence on the Dropwizard website:

> Dropwizard pulls together stable, mature libraries from the Java ecosystem into a simple, light-weight package that lets you focus on getting things done.

It's not the goal of Dropwizard to be as modular and as exchangeable as possible, but to provide a well-integrated set of components which users can immediately start working with.

So far, Jetty has proven to be a good choice as the only embedded container supported by Dropwizard.

**Comment by cricket007** [comment]
Thanks for the responses, and I value the opinions that have been made to make each library work so well together out of the box with a minimal configuration; I'm simply trying to start a discussion, not truly argue over performance or chosen/supported features. 

I understand benchmarking must be performed on one's own infrastructure to truly identify bottlenecks, however I'd argue not everyone has the time and resources to do so. 

I'd like to think the reason Spring has become so popular is the <strike>"market branding"</strike> word of mouth, but that had to start *somewhere*, and likely the conversation was "hey, I've used this, and it works for me", then only later did more integrations get added to it, then Netflix OSS pushed their code into it, after Spring Boot 2 became a large improvement, etc, etc. I'm totally in agreement it's bloated. (but the Spring Initalizer site is really cool for getting all dependencies aligned, and even Vertx & Quarkus basically forked that initializer page).   
I'm not trying to get you to add feature parity anywhere close to Spring ecosystem, only providing different HTTP (or even other protocol} handlers. 

> Dropwizard is opinionated and it should stay opinionated.

I accept Jetty as the accepted framework, and JAX-RS works fine. People new to web services or coming from most other Java web applications should be able to easily grasp it, as compared to Akka actors or Vertx verticles. My personal opinion says that it could possibly be more extensible, with the default being still Jetty. Similar to how Spring's default is Tomcat (which in my opinion is worse than Jetty *or Undertow*, but hey, Spring devs are entitled to their own opinions too) 

I (and OP) are simply looking for guidance if Dropwizard even *could* be less coupled to it. One use case I can think of would be wrapping libraries written using other embedded servers. In the case of my contribution to Armeria, it offered Thrift over gRPC services rather than Protobuf like most gRPC frameworks do. Plus, it seems to document the asynchronous framework aspects of it than Dropwizards documentation does for Jersey (even if code examples were copied straight out of the Jersey documentation).   
I realize there's already Dropwizard gRPC and Protobuf bundles out there, but even those are external, like I am suggesting these other embedded servers could be. 

Now, you might say, "why should it be the responsibility of Dropwizard to cater to those familiar with other containers?" and I say because you're ultimately "losing 'customers'." Either by pushing them away by saying that you cannot offer developers an onboarding experience, or forcing them to rewrite their applications rather than welcoming them to the community or supporting their interest in Dropwizard. Alternatively, one could try to convince the Undertow or other communities to add a module for Dropwizard, and put the responsibility on them (like I did for Armeria). Either way, I understand there would be a development burden with little to no gain. 

> Adding support for more servers means more code, consequently more bugs, worse out-of-the-box configuration and adds up a lot to the maintenance, adding no value.

I'm not saying the core maintainers (you, et al) need to even "maintain" these containers, just make the ServerFactory, Environment, Lifecycle, and whatever else from core more extensible by abstracting away the explicitly Jetty usages. From what I understand, isn't this the reason `dropwizard-jetty` is it's own module? 

Why do the server factory options have to be limited to basic or simple? 

> tl;dr I don't think that swapping Jetty with Undertow in Dropwizard will provide any gains

Probably not, but I prefer composition of libraries and frameworks rather than having to port or deploy them side by side, as one would currently have to do if migrating or using something based on Undertow or other embedded servers. 

If one really wanted performance, I would think it'd be Vertx based, which is what RedHat chose for Quarkus (which also supports JAX-RS based resources) rather than their JBoss / Wildfly / Thorntail / whatever other libraries they back & develop. 

> Dropwizard (71), which is basically Jetty + Jersey, is faster than Undertow+Jetty (74)

You have a point there. I missed that undertow+jersey one. 

I'll point out that Armeria is 63, though, so I plan on contributing armeria-dropwizard to that benchmark's Github page for their next round.

In addition, they seem to currently be on Dropwizard 1.3.12 or 15, IIRC, so I plan on contributing there anyway to see if that makes any improvement. 

> are those benchmarks relevant to any real world scenario, not at all.

I'll accept that criticism, but you can't assume people won't search for "best/fastest Java web framework"

Spring is the kitchen sink. Vertx is fast. Micronaut is targeting the cloud native Spring-like experience at a smaller footprint. Helidon wants to support both JavaEE and JavaSE workflows. Quarkus likes advertising their heap usage and start up speeds plus GraalVM support. Spark, Play, and Akka HTTP are aimed at the Functional devs, yadayada. 

 There's too many options out there, yet Dropwizard very rarely gets the credit it deserves, in my opinion. Maybe lack of blogs, examples, "preaching to the choirs"? I don't know. 

But I want to end this by saying, thank you for all your time and efforts in developing such a great framework, and providing your feedback and opinions. If you're not willing to consider our feature requests, then so be it. Regardless, happy new year to you all, and hope Dropwizard 2.x+ brings lots of more useful features to the framework.
